### PR TITLE
Fl initial

### PR DESCRIPTION
**Pull request by viktorvaladi**
Initial FL code.

### CODE REVIEW

File: leakpro/user_inputs/cifar10_gia_input_handler.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+    def get_criterion(self:Self)->None:
+        """Set the CrossEntropyLoss for the model."""
+        return torch.nn.CrossEntropyLoss()
+
+    def get_optimizer(self: Self) -> MetaOptimizer:
+        """Set the optimizer for the model."""
+        optimizer = self.configs["audit"]["gia_settings"]["optimizer"]
+        lr = self.configs["audit"]["gia_settings"]["learning_rate"]
+        if optimizer == "SGD":
```
Comment (johanos1) [inline leakpro/user_inputs/cifar10_gia_input_handler.py:33]: This mapping should be in the abstract class, the user should not have to know about this.

File: leakpro/user_inputs/cifar10_gia_input_handler.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+        Requires a meta optimizer that performs step to a new set of parameters to keep a functioning
+        graph.
+
+        Training does not update the original model, but returns a norm of what the update would have been.
+        """
+
+        gpu_or_cpu = device("cuda" if cuda.is_available() else "cpu")
+        self.target_model.to(gpu_or_cpu)
+        patched_model = MetaModule(deepcopy(self.target_model))
```
Comment (johanos1) [inline leakpro/user_inputs/cifar10_gia_input_handler.py:57]: do we need to make it a deepcopy?

Comment (johanos1) [reply leakpro/user_inputs/cifar10_gia_input_handler.py:57]: It would be nice if we could make this invisible to the user, i.e., the user shouldnt have to know we have this patched model. Any ideas?

File: leakpro/attacks/gia_attacks/invertinggradients.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+
+class InvertingGradients(AbstractGIA):
+    """Gradient inversion attack by Geiping et al."""
+
+    def __init__(self: Self, handler: AbstractGIAInputHandler, configs: dict) -> None:
+        super().__init__(handler)
+        self.handler = handler
+        self.logger.info("Inverting gradient initialized :)")
+        self.t_v_scale = configs.get("total_variation")
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/invertinggradients.py:18]: Please add a function to ensure these parameters are within correct ranges, see MIA attacks.

File: leakpro/attacks/gia_attacks/invertinggradients.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        self.attack_lr = configs.get("attack_lr")
+
+    def description(self:Self) -> dict:
+        """Return a description of the attack."""
+        title_str = "Inverting gradients"
+        reference_str = "Geiping, Jonas, et al. Inverting gradients-how easy is it to break privacy in federated learning?(2020)."
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/invertinggradients.py:24]: Neurips, 2020

File: leakpro/attacks/gia_attacks/invertinggradients.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        -------
+            None
+
+        """
+        self.handler.target_model.eval()
+        self.client_loader = self.handler.client_data
+        self.data_mean, self.data_std = self.handler.get_meanstd()
+        self.reconstruction = self.handler.init_at_image()
+
+        # Extract the labels from self.handler.client_data
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/invertinggradients.py:51]: State that this is an assumption we make, i.e., the label distribution is known.

File: leakpro/attacks/gia_attacks/invertinggradients.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        labels = []
+        for _, label in self.handler.client_data:
+            labels.extend(label.numpy())
+        labels = torch.tensor(labels)
+        reconstruction_dataset = TensorDataset(self.reconstruction, labels)
+        self.reconstruction_loader = DataLoader(reconstruction_dataset, batch_size=32, shuffle=True)
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/invertinggradients.py:57]: why is batch size hard-coded? Also, should it really be shuffled?

File: leakpro/attacks/gia_attacks/invertinggradients.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        """Run the attack and return the combined metric result.
+
+        Returns
+        -------
+            GIAResults: Container for results on GIA attacks.
+
+        """
+        self.reconstruction.requires_grad = True
+        optimizer = torch.optim.Adam([self.reconstruction], lr=self.attack_lr)
+        scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer,
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/invertinggradients.py:72]: Is this required?

File: leakpro/attacks/gia_attacks/invertinggradients.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+                                                            milestones=[24000 // 2.667, 24000 // 1.6,
+
+                                                                        24000 // 1.142], gamma=0.1)
+
+        for i in range(500):
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/invertinggradients.py:77]: 500 should be a parameter read from config.

File: leakpro/attacks/gia_attacks/invertinggradients.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+            closure = self.gradient_closure(optimizer)
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/invertinggradients.py:78]: comment please

File: leakpro/attacks/gia_attacks/invertinggradients.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+        for i in range(len(self.client_loader.dataset)):
+            x, _ = self.client_loader.dataset[i]
+            all_data.append(x)
+
+        # Convert lists to tensors
+        original_data_tensor = torch.stack(all_data)
+        return GIAResults(original_data_tensor, self.reconstruction, 0, self.data_mean, self.data_std)
+
+    def total_variation(self: Self, x: torch.Tensor) -> torch.Tensor:
+        """Anisotropic TV."""
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/invertinggradients.py:101]: This should be in util

File: leakpro/attacks/gia_attacks/invertinggradients.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+            optimizer.zero_grad()
+            self.handler.target_model.zero_grad()
+
+            gradient = self.handler.train(self.reconstruction_loader, self.handler.get_optimizer())
+            rec_loss = self.reconstruction_costs([gradient], self.client_gradient)
+
+            self.logger.info(f"rec loss {rec_loss}")
+            rec_loss += self.t_v_scale * self.total_variation(self.reconstruction)
+            rec_loss.backward()
+            self.reconstruction.grad.sign_()
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/invertinggradients.py:135]: what does this do?

File: leakpro/attacks/gia_attacks/invertinggradients.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+            Returns
+            -------
+                torch.Tensor: The reconstruction loss.
+
+            """
+            optimizer.zero_grad()
+            self.handler.target_model.zero_grad()
+
+            gradient = self.handler.train(self.reconstruction_loader, self.handler.get_optimizer())
+            rec_loss = self.reconstruction_costs([gradient], self.client_gradient)
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/invertinggradients.py:130]: do we need this to be list?

File: leakpro/attacks/gia_attacks/invertinggradients.py [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+
+            self.logger.info(f"rec loss {rec_loss}")
+            rec_loss += self.t_v_scale * self.total_variation(self.reconstruction)
+            rec_loss.backward()
+            self.reconstruction.grad.sign_()
+            return rec_loss
+        return closure
+
+    def reconstruction_costs(self: Self, client_gradients: torch.Tensor, reconstruction_gradient: torch.Tensor) -> torch.Tensor:
```
Comment (johanos1) [inline leakpro/attacks/gia_attacks/invertinggradients.py:139]: we only consider a single client gradient?

**Review by johanos1 — CHANGES_REQUESTED** [review]

**Comment by johanos1** [comment]
Nice! Almost there. :)

File: leakpro/user_inputs/abstract_gia_input_handler.py [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+class AbstractGIAInputHandler(ABC):
+    """Parent class for user inputs."""
+
+    def __init__(self:Self, configs: dict, logger:logging.Logger, client_data: DataLoader, target_model: torch.nn.Module) -> None:
+        self.configs = configs
+        self.logger = logger
+        self.client_data = client_data
+        self.target_model = target_model
+
+    def init_at_image(self: Self) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (viktorvaladi) [reply leakpro/user_inputs/abstract_gia_input_handler.py]: It's moved to data_module now.

I believe it is attack agnostic so better to keep it out of the attack. The new implementation will create a dataloader with random noise images in the shape of a client loader. Later we can fill in with other types of initialization that people have investigated, e.g. the one from nvflare paper.

File: leakpro/user_inputs/abstract_gia_input_handler.py [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+        """Initializes images with random pixels."""
+        img_shape = self.client_data.dataset[0][0].shape
+        num_images = len(self.client_data.dataset)
+        return torch.randn((num_images, *img_shape),
+                           **{"device": next(self.target_model.parameters()).device,
+                               "dtype": next(self.target_model.parameters()).dtype}
+                               )
+
+    def get_meanstd(self: Self) -> Tuple[list, list]:
```
[the remark this answers is not in the document]
Comment (viktorvaladi) [reply leakpro/user_inputs/abstract_gia_input_handler.py]: Fixed now the method is contained in the data_module.

File: leakpro/attacks/gia_attacks/invertinggradients.py (hunk 7 above)
Comment (viktorvaladi) [reply leakpro/attacks/gia_attacks/invertinggradients.py:64]: yes, torch optimizers expects something iterable thats why we put it in a list.

File: leakpro/attacks/gia_attacks/invertinggradients.py (hunk 11 above)
Comment (viktorvaladi) [reply leakpro/attacks/gia_attacks/invertinggradients.py:119]: added a comment:
"# Add the TV loss term to penalize large variations between pixels, encouraging smoother images."

**Review by johanos1 — APPROVED** [review]
